### ISSUE TITLE
More sophisticated/customizable error prevention methods

### ISSUE THREAD

**Issue by noctuid** [keyboardio/Kaleidoscope-DualUse#7]
I don't think that a timeout is a sufficient method for intelligent dual-role. I have a personal software dual-role implementation that I created for testing that allows for any combination of the following methods:

1. customizable timer for how long dual-role key has been held (how dualuse currently works)
2. customizable timer for minimum required overlap between between dual-role key and other keys (keys generally have less overlap time when typing than when using a modifier combination; I've found somewhere around 50ms to be sufficient to prevent a lot of errors using this method alone)
3. checking release order (#1)

To consistently prevent errors when using the timeout method, the timeout has to be fairly long, which makes the dual-role key tedious to use. I was able to set the timeout to closer to 100-150ms using other methods in combination with the timeout which makes the dual-role key a lot more natural to use.

Like you pointed out in #1, requiring the other key to be released first breaks holding a key for repeating, but I'm not suggesting this method as an alternative but as an optional addition. Ideally all of these methods could be chosen/customized on a per-key basis. There shouldn't be any need for a tradeoff; just have only the timeout be the default and let the user decide if they want to use the other methods.

Even just allowing the timeout to be set on a per-key basis would be useful. Long timeouts are only necessary to prevent mistakes for non-modifier keys that are typed in rapid succession with other keys (space, letters keys, etc.). I'm fine having zero timeout for keys like escape, enter, tab, etc.

With this change, `MT` would look something like `MT(mod, key, timeout, minimum overlap time, check on release?)` (last 3 optional and global defaults settable for the timers).

**Comment by gedankenexperimenter** [comment]
I've been thinking about this again, and I'm back to something like my original position (see #1) — timers alone are inadequate for DualUse keys that are meant to be used for regular typing (e.g. letter keys).

I hadn't thought of requiring a minimum overlap; I'm still focused on how to do it properly by key release order, and was thinking a short delay (100ms?) after a DualUse key is pressed and held before registering the modifier (for use with other input devices). Then using release order to decide what to do if a second key is pressed within that window. Overlap might be a better approach, but my feeling is that it should only matter if the DualUse key is released first (i.e. ambiguous situations).

**Comment by noctuid** [comment]
I'm not sure I understand. Are you suggesting a combination of timeout and checking release order where the release order is only checked if the second key is also pressed with some time interval? If I'm understanding correctly, I don't think this would help. For myself and probably most people, when using the dual-role key as a modifier, it will always be released last. On the other hand, when testing with my own program, I found that the dual-role key still sometimes incorrectly acted as a modifier during typing when combining a short timeout with checking key release order (when I type, keys are not always released in the order they are pressed). It may be possible to lower this error rate by further increasing the timeout (and using a mechanical keyboard probably helps), but this is undesirable, and I found that enforcing a minimum overlap time acts as a safety net without downsides.

My previous suggestions are based on these ideas:
- The user shouldn't have to adjust their typing; they shouldn't have to consciously hold down modifiers longer or not roll between keys when typing (I know of someone who is able to succesfuly use space as a modifier without any timeout or advanced checking by not ever "rolling" to the spacebar when typing; this requires retraining and is a lot less doable with letter keys)
- There are distinct characteristics of modifier usage that can be recognized; modifiers are held longer (hence timeout), stay depressed for the entire duration of other keys pressed with them (hence overlap checking), and are released last (hence release order checking)

If someone happened to use modifiers more like chords (i.e. very quickly pressing and releasing the combo or releasing the modifier before the other key(s)), these assumptions wouldn't be true. This is one reason I think each method should be optional. I think the most likely scenario where a user wanted the key to act as a modifier and it didn't would be if they released it just before the other key pressed with it. If this was an issue, the user could just use the timeout and overlap methods.

I should also mention that I check to see if the timeout has been reached on key release and think this is how it should be done when timeout and release order are both enabled. This allows further taking advantage of the fact that the modifier will still be held while the other key is pressed. When checking the timeout as soon as another key is pressed with it, a very short timeout is required if the user doesn't want to intentionally wait for the key to start acting as a modifier. Checking on release allows the user to have a longer timeout (making errors less likely), without having to wait.
